File: src/main/java/checkdigitcli/algorithm/Algorithms.java
package checkdigitcli.algorithm;

import java.util.HashMap;
import java.util.Map;

import checkdigitcli.main.CheckDigitCLIException;

public class Algorithms {

	public static final Map<String, Algorithm> ALGORITHMS;

	static {
		ALGORITHMS = new HashMap<>();
		
		Luhn luhn = new Luhn();
		ALGORITHMS.put(luhn.getName(), luhn);
	}

	public static Algorithm getAlgorithm(String algorithmName) throws CheckDigitCLIException {

		// Check for algorithm internally
		if (Algorithms.ALGORITHMS.containsKey(algorithmName)) {
			Algorithm a = Algorithms.ALGORITHMS.get(algorithmName);
			System.out.println("Using algorithm " + a.getClass().getSimpleName());
			return a;
		}

		throw new CheckDigitCLIException("No algorithm with name " + algorithmName);
	}
}

File: src/main/java/checkdigitcli/algorithm/Algorithms.kt
package checkdigitcli.algorithm

import checkdigitcli.main.CheckDigitCLIException

object Algorithms {
    private val ALGORITHMS: MutableMap<String, Algorithm> = HashMap()

    init {

        val luhn = Luhn()
        ALGORITHMS[luhn.name] = luhn
    }

    @Throws(CheckDigitCLIException::class)
    fun getAlgorithm(algorithmName: String): Algorithm {

        // Check for algorithm internally
        if (ALGORITHMS.containsKey(algorithmName)) {
            val a = ALGORITHMS[algorithmName]
            println("Using algorithm " + a!!.javaClass.simpleName)
            return a
        }
        throw CheckDigitCLIException("No algorithm with name $algorithmName")
    }
}